mlflow/pyfunc/__init__.py

```diff
@@ -1648,12 +1662,160 @@ def _convert_struct_values(
+        raise MlflowException(
+            f"The prebuilt env '{env_archive_path}' runtime version '{prebuilt_runtime_version}' "
+            f"does not match UDF sandbox runtime version {runtime_version}."
+        )
+    if prebuilt_runtime_version != runtime_version:
+        raise MlflowException(
+            f"The prebuilt env '{env_archive_path}' platform machine '{prebuilt_platform_machine}' "
+            f"does not match UDF sandbox platform machine {prebuilt_platform_machine}."
+        )
+
+
+def _prebuild_env_internal(local_model_path, archive_name, save_path):
+    env_root_dir = os.path.join(_PREBUILD_ENV_ROOT_LOCATION, archive_name)
+    archive_path = os.path.join(save_path, archive_name + ".tar.gz")
+    if os.path.exists(env_root_dir):
+        shutil.rmtree(env_root_dir)
+    if os.path.exists(archive_path):
+        os.remove(archive_path)
+
     try:
-        from pyspark.sql.utils import is_remote
-    except ImportError:
-        return False
-    return is_remote()
+        pyfunc_backend = get_flavor_backend(
+            local_model_path,
+            env_manager="virtualenv",
+            install_mlflow=False,
+            create_env_root_dir=False,
+            env_root_dir=env_root_dir,
+        )
+
+        pyfunc_backend.prepare_env(model_uri=local_model_path, capture_output=False)
+        # exclude pip cache from the archive file.
+        shutil.rmtree(os.path.join(env_root_dir, "pip_cache_pkgs"))
+
+        return archive_directory(env_root_dir, archive_path)
+    finally:
+        shutil.rmtree(env_root_dir, ignore_errors=True)
+
+
+def prebuild_model_env(model_uri, save_path):
+    """
+    Prebuild model python environment and generate an archive file saved to provided
+    `save_path`.
+
+    Args:
+        model_uri: URI to the model that is used to build the python environment.
+        save_path: The directory path that is used to save the prebuilt model environment
+            archive file path.
+
+    NOTE: The `prebuild_model_env` can only be called in Databricks runtime,
+        and the generated prebuilt env file can be used in `mlflow.pyfunc.spark_udf` in
+        Databricks runtime or Databricks Connect client.
+        The prebuilt env archive file can't be used across different Databricks runtime
+        versions or different platform machines.
+        When using the prebuilt env in `mlflow.pyfunc.spark_udf`, MLflow will verify
+        whether the spark UDF sandbox environment matches the prebuilt env requirements.
+    """
+    from mlflow.utils._spark_utils import _get_active_spark_session
+
+    if not is_in_databricks_runtime():
+        raise RuntimeError("'prebuild_model_env' only support running in Databricks runtime.")
+
+    if not os.path.exists(save_path):
+        raise RuntimeError(f"The saving path '{save_path}' does not exist.")
+    if not os.path.isdir(save_path):
+        raise RuntimeError(f"The saving path '{save_path}' must be a directory.")
```

Comment:
can `save_path` contain files?

Reply:
yes.

Reply:
I see, so it doesn't have to be an empty?

Reply:
User can use the generated *.tar.gz file anywhere, the parent directory doesn't matter.